Add DCP/2 System Design (#2)
HumanCellAtlas/dcp2: docs/dcp2_system_design.rst

                "output_type": "analysis_file",
                "output_id": "2e072336-b906-4c9b-a475-ca03f51c8452"
            }
        ],
        "protocols": [
            {
                "protocol_type": "analysis_protocol",
                "protocol_id": "4d6f7580-ce81-4a81-9c2c-872fcb23b7cd"
            }
        ]
    }

The steps to reconstruct a subgraph from the tables in a TDR *dataset*
are as follows:

#. Given a ``links_id`` (the UUID of a subgraph) and
   ``version``, fetch the row with that ``links_id``
   and ``version`` from the ``links`` table.

#. Read that row's ``content`` column and iterate over the items
   of the ``links`` property (a list) of the JSON document in
   ``content``. For each item aka link,

   a. read the ``process`` property, and extract the schema type
      from the ``process_type`` property and the process
      UUID from the ``process_id``. Query the TDR table that
      corresponds to the schema type and fetch all rows where
      ``{schema_type}_id`` equals the process UUID. Pick the row
      with the highest version.

   b. read the ``inputs``, ``outputs`` and
      ``protocols`` properties (they're all lists). For each

>       input, output and protocol,

Reviewer: (nitpick) These sub-bullets display as `i.` and `ii.` instead of `a.` and `b.` (Original doc display these as `a.` and `b.`)

Author: This seems to be determined by the rendering engine (Github in this case). I thought the manual numbering would come through, but since it didn't, I changed it all to auto-numbering.

subgraphs, but also the need to learn BigQuery and the TDR schema.«

Supplementary files
-------------------

… in DCP/1 were not linked as part of ``links.json``. Their mere presence
in a bundle associated them with the project. There are no bundles in
DCP/2 so, the schema for ``links.json`` has been refactored to
accommodate a new type of ``supplementary_file_link``, in
addition to the existing ``process_link``. This new type of link
is used to associate an entity (currently only those of type
`project`_) with a `supplementary_file`_.
The DSS adapter scans each bundle for supplementary files and adds links
to ``link.json`` accordingly.

.. _project: https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/type/project/project.json
.. _supplementary_file: https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/type/file/supplementary_file.json

Generic file metadata
---------------------

The bundle manifest entries in the DCP/1 datastore (DSS) contain certain
data file properties that aren't captured anywhere else. Because of the
absence of bundles in DCP/2, these properties are instead stored in a

> *file descriptor*, as defined by the `file_descriptor.json`_

Reviewer: (nitpick) `file_descriptor.json` displays as a link but not as ``code`` block, perhaps not possible to do both?

Author: Essentially no. See https://docutils.sourceforge.io/FAQ.html#is-nested-inline-markup-possible

   quality comparison

Analysis provenance
-------------------

In DCP/1, an analysis bundle (a bundle containing output files from an
analysis workflow) referred to the input bundle (a bundle that contains
the input files) via the
`input_bundles`_ property of the ``analysis_process`` entity. [6]_ This was
problematic in two ways: 1) the bundle version is missing and 2)
metadata should be agnostic to bundles. The ``analysis_process``
schema also requires a ``reference_bundle`` property for
specifying the bundle that contains the reference files. This property
also suffers from the same problems. The two properties of an
``analysis_process`` entity and the ``links.json`` files
are the only places where metadata mentions bundles in DCP/1.

.. _input_bundles: https://github.com/HumanCellAtlas/metadata-schema/blob/f37da4858d0a31263d2126246e552f45048cb87c/json_schema/type/process/analysis/analysis_process.json#L184

To solve both problems, the ``analysis_process`` schema was
revised to instead list references to the metadata entities that
describe the individual reference data files in a property called
``reference_files``. The ``input_bundles`` and

> ``reference_bundle`` properties was removed. Note that an entry

Reviewer: ```suggestion
``reference_bundle`` properties were removed. Note that an entry
```
(grammar nitpick of original doc)

   are available when listing datasets and snapshots using the TDR
   API. Additionally, IDs are hard to read to the human eye, and
   hard to distinguish visually, so as long as we manually confer
   them between teams, names are preferred.«

Grammars below use a mix of `EBNF`_
and regular expressions. They are designed to be easily parsed, either
using regular expressions or by splitting on underscore, and so that a
lexicographical sorting reflects both the hierarchical relationship
between datasets and snapshots as well as the time they were created. ::

    dataset_name = "hca\_" , deployment , "_" , creation_date , ["_" ,
    qualifier]

    snapshot_name = dataset_name , "_", [project_id], "__" , creation_date ,
    ["_" , qualifier]

.. _EBNF: https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form

Note that snapshots have two dates in their name. The three consecutive
underscores are not a typo, they exist to facilitate snapshot tiling for
implementing controlled access in the future. ::

    creation_date = year , month , day

    year = [0-9]{4}

    month = [0-9]{2}

    day = [0-9]{2}

    qualifier = [a-zA-Z][a-zA-Z0-9]{0,13}

>     deployment = "dev" \| "staging" \| "prod"

Reviewer: ```suggestion
    deployment = "dev" | "staging" | "prod"
```

    project_id = [0-9a-f]{8}(-[0-9a-f]{4}){3}-[0-9a-f]{12}

The following regex should be used to validate dataset names::

    ^hca_(dev|prod|staging)_(\d{4})(\d{2})(\d{2})(_[a-zA-Z][a-zA-Z0-9]{0,13})?$

To validate snapshots (line break added for legibility)::

    ^hca_(dev|prod|staging)_(\d{4})(\d{2})(\d{2})(_[a-zA-Z][a-zA-Z0-9]{0,13})?

>     \___(\d{4})(\d{2})(\d{2})(_[a-zA-Z][a-zA-Z0-9]*)?$

Reviewer: ```suggestion
    ___(\d{4})(\d{2})(\d{2})(_[a-zA-Z][a-zA-Z0-9]*)?$
```

Staging area layout
-------------------

Object names given in this section are relative to the staging area. To
produce the complete ``gs://…`` URI of a particular object in the
staging area, append the object's name to the staging area URI. [7]_

There are four object naming schemes, one for data files, one for file
descriptors, one for metadata files and one for ``links.json`` files.

-  The object name of a metadata entity is::

        metadata/{entity_type}/{entity_id}_{version}.json

   where

   ``entity_type``
       is the `HCA schema entity type`_
       such as ``cell_suspension``.

   ``entity_id``
       is a UUID that uniquely identifies the metadata
       entity. [8]_ The TDR importer uses ``entity_id`` as the
       PK for the row in the corresponding BQ table e.g., the
       ``cell_suspension_id`` column of the
       ``cell_suspension`` table.

   ``version``
       is an ISO timestamp with the ``Z`` suffix
       for UTC and a six-digit number of microseconds left-padded with
       leading zeros if necessary, e.g.,
       ``2020-05-01T04:26:07.021870Z``. Not every ISO syntax is
       supported, only those that match the regex:

>        ``\\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}Z`` To avoid

Reviewer: ```suggestion
       ``\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}Z`` To avoid
```

       ambiguous string representations of the same version timestamp,
       the ``Z`` suffix is mandated and the microseconds field
       can't be omitted. Whole seconds must be specified as ``.000000``.
       The DSS adapter uses the DSS version of the metadata file,
       converting it to this restricted ISO syntax.

.. _HCA schema entity type: https://github.com/HumanCellAtlas/metadata-schema/tree/master/json_schema/type

-  The object name of a file descriptor is::

        descriptors/{entity_type}/{entity_id}_{version}.json

   where ``entity_type``, ``entity_id`` and
   ``version`` have the same meaning as for metadata entities,
   except that the value of ``entity_type`` has to end in

>    ``\_file``. File descriptors are JSON documents and are

Reviewer: ```suggestion
   ``_file``. File descriptors are JSON documents and are
```

   described in `Generic file metadata`_.

-  The object name of a data file is::

       data/{file_name}

   where

   ``file_name``
       is the ``file_name`` property from the
       file descriptor object for this data file.

-  The object name for ``links.json`` files is::

       links/{links_id}_{version}_{project_id}.json

   where

   ``links_id``
       is a UUID that uniquely identifies the subgraph.
       The DSS adapter uses the bundle UUID.

   ``version``
       field uses the same syntax as in object names
       for metadata entities. It denotes the version of the subgraph.
       The DSS adapter uses the DSS bundle version.

   ``project_id``
       field identifies the project the subgraph
       is part of. A subgraph is part of exactly one project. The
       ``project_id`` field is at the end of the object path so
       that the importer, using a prefix query, can look up subgraph
       objects by their ID without knowing the project ID. The importer
       must record an error if it detects more than one object with the

>        same ``links/{links_id}_{version}\_`` prefix.

Reviewer: ```suggestion
       same ``links/{links_id}_{version}_`` prefix.
```

``{links_id}/{content.file_core.file_name}``).«

Initially, the bundle-relative path of a data file in DCP/1 was not
allowed to contain slashes, forcing bundles to be flat. This limitation
has been in part addressed, but when the DCP/1 was shut down, many
bundles (like those with zarray matrices) still contained file names
that work around this limitation by using ``!`` instead of ``/``. The DSS
adapter undoes the substitution. That is the only modification the DSS
adapter applies to the `file_core.file_name`_ property.

Importer idempotence
--------------------

The importer should not copy a data file if it is already present in TDR
and the checksums match between the copy of the file in the staging area
and the one in TDR/Firestore.

Similarly, the importer must not create a new row in a TDR table if that
row would be identical to another row except for its ``rowid``.

Summary of schema changes
-------------------------

»This section lists the schema changes that were made as part of the MVP
release of the DCP/2. The table is for information only.«

.. list-table::

    * - file_descriptor.json
      - `metadata-schema #1289`_
        `metadata-schema #1302`_

>         `metadata-schema #1317`_

Reviewer: These links display space separated. Is there any way to make them display newline separated like in the original doc?

Author: Yes, great suggestion.

Invalid documents in the DCP/1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There was one known existing schema violation in documents in the DCP/1
production instance of DSS. The
``provenance.schema_major_version`` and
``provenance.schema_minor_version`` properties are present in
metadata files submitted to the DSS from around Oct 2019 onwards. The
addition of these fields was proposed in `RFC 11`_.
After that RFC was accepted, the ``provenance`` schema was
revised and the Ingest component was modified to add those fields to
documents but the schema reference in those documents still points at an
old schema, not the revised one. This is further complicated by the fact
that the ``provenance`` schema is referenced indirectly via the
main document schema (`example document`_).
The ``analysis_file`` schema in that document is at version
6.0.0 while the new fields were introduced in version 6.2.0 (via the
``provenance`` schema version 1.1.0). The problem affects all
metadata documents submitted after October of 2019, not just
``analysis_file`` documents.

.. _RFC 11: https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0011-query-by-metadata-schema-versions.md

> .. _example document: http://dss.data.humancellatlas.org/v1/files/003d3dda-6906-4943-9f12-331b963e2f55?replica=aws

Reviewer: Link URL matches original doc but is a broken link. Is there a replacement?

Author: I'll file a new issue from this to be solved later ;)

Author: #3

To address this issue, the DSS adapter removes those two fields and the
Ingest adapter is modified to not emit them. Luckily, the fields are not
required so removing them from documents that **do** happen to carry an
updated schema declaration does **not** invalidate those documents. `The
provenance schema will be revised to remove the fields
again <provenance schema_>`_.

.. _provenance schema: https://github.com/HumanCellAtlas/metadata-schema/issues/1316

It was also determined that some metadata documents in DCP/1 contain
schema URLs with a host name that points to a non-production instance of
the site where the schemas are published, namly
``schema.staging.data.humancellatlas.org`` and
``schema.dev.data.humancellatlas.org``. It was established that
the schemas referenced via these URLs were identical to their
counterparts on ``schema.humancellatlas.org``. In order to keep
the affected documents valid, the DCP/2 will maintain the non-production
instances until the metadata documents are corrected and we stop
supporting the releases affected by this issue (currently dcp1 and
dcp2).

Import errors
-------------

> This work is currenly in progress and tracked by DSPDC-1604.

Reviewer: Line is not present in original doc.

Author: @hannes-ucsc requested I add this. Nice catch though.

Errors that occur during the importer's processing of the staging area
for a particular source are logged by the importer to dedicated files in
the staging area. The object naming scheme for these error log files is
``errors/{timestamp}.json`` where ``timestamp`` is the
start time of the importer invocation. The format of the file is

> `JSONL`_ but only errors should be logged. An

Reviewer: ```suggestion
`JSON Lines`_ but only errors should be logged. An
```
(nitpick of original doc) `.jsonl` is the file extension but `JSON Lines` is the format.

https://jsonlines.org/
> This page describes the JSON Lines text format, also called newline-delimited JSON

empty file indicates that no errors occurred.

.. _JSONL: http://jsonlines.org/

The format of the error message will be as follows::

>     {
>     "errorType": "string",
>     "filePath": "string",
>     "fileName": "string",
>     "message": "string"
>     }

Reviewer: ```suggestion
    {
        "errorType": "string",
        "filePath": "string",
        "fileName": "string",
        "message": "string"
    }
```

For example::

    {
        "errorType": "SchemaValidationError",
        "filePath": "metadata/organoid/123456_VERSION1.json",
        "fileName": "123456_VERSION1.json",

>         "message": "Data does not conform to schema from
>                     https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/type/biomaterial/organoid.json;
>                     missing required field ‘describedBy’”

Reviewer: ```suggestion
        "message": "Data does not conform to schema from https://github.com/HumanCellAtlas/metadata-schema/blob/master/json_schema/type/biomaterial/organoid.json; missing required field ‘describedBy’”
```
 Technically not the same "message" value due to the spaces for the indentation being added inside the quotes.

Author: Good catch, I'll change this back.

holding the `path of the corresponding data file <Data file paths_>`_
must not be modified and the data file object should be omitted from the
staging area. The TDR importer detects this case and modifies the
corresponding row in the TDR table but reuses the existing data file in
Firestore [I might be using the wrong terminology here]. This is a
direct consequence of the `importer's idempotence <Importer idempotence_>`_.
If the referenced data file does not exist in Firestore, the importer
`records an error <Import errors_>`_. There is a separate section
[TODO: reference section] dedicated to data files updates.

After the source applies this process to all entities it intends to
update, it requests the importer to be invoked.

Notes from comments:

-  Cases for DELETE markers:

   -  Impractical to stage Tabula Muris except one object only to delete
      that object.

   -  Deletions should be explicit actions, rather than a potentially
      accidental absence of an object.

-  Delta staging area partiality:

   a) omit any object that hasn't change

   b) (stricter) require that subgraphs are fully defined

   c) like b but allow omission of data files

>    d) like c but also allow omission of descriptors

Reviewer: `a)` `b)` `c)` `d)` display as `i.` `ii.` `iii.` `iv.` making the references `like b` and `like c` invalid.

Author: I can't figure out how to control the numbering style; I don't think it's possible. I changed these to auto-number and changed the text to say "like previous, but..."

an adequate metadata schema by which to describe these matrices.
Reproducibility of analysis that uses the interim artefacts will be
provided via TDR snapshots. The MVP snapshot will permanently expose the
interim artifacts, while a later release snapshot will expose the
permanent replacements.

For DCP/2 MVP, no post-processing is performed on contributor-generated
matrices. Users will be able to download them in exactly the same file
format as they were provided by the contributor. For example, if the
contributor provided a ZIP archive of multiple matrix files, the user
will need to download that archive and extract it locally.

The interim solution stores the contributor-generated matrices as data
files in the TDR and describes them with a `file
descriptor <Generic file metadata_>`_, a ``supplementary_file`` entity
and a ``supplementary_link`` entry in a ``links.json``.
There will be one subgraph (and therefore one ``links.json``
document) per HCA project. If the contributor provided multiple matrices
stratified by, say, species and organ, the per-project subgraph will
contain a ``supplementary_file`` and a
``supplementary_link`` for each such matrix file.

> .. image:: media/project-cgm.png

Reviewer: (nitpick) can the image be centered?

Reviewer: Markup added to center the image didn't seem to carry over to the displayed document. Not a major issue however.
![Screen Shot 2021-03-10 at 11 02 35 AM](https://user-images.githubusercontent.com/46407464/110682742-23278b80-8190-11eb-8d38-297f09093fcd.png)

of a ``supplementary_file`` entity, the stratification of a
DCP/2-generated matrix is determined organically by traversing the
metadata graph.

The DCP/2 analysis team assembles stratified project-level matrices from
smaller intermediate matrices. The project-level matrix will be staged
in a separate subgraph that links the project-level matrix to the
intermediate matrices. The subgraphs containing the intermediate
matrices link to the input sequence files which are in separate primary
subgraphs. Azul recursively stitches these three levels of subgraphs
together.

Project level matrix files are described by an ``analysis_file`` entity
with the following properties:

-  The ``analysis_file.provenance.submitter_id`` field is set to
   ``e67aaabe-93ea-564a-aa66-31bc0857b707``. This UUIDv5 was
   generated using the same namespace UUID as for
   `Contributor-generated matrices (CGMs)`_ and the name
   ``dcp2``.

-  The ``analysis_file.file_core.content_description`` field of both
   intermediate and project-level matrices is set to::

       {
           "text": "DCP/2-generated matrix",
           "ontology": "data:3917",
           "ontology_label": "Count Matrix"
       }

> -  The ``analysis_file`` entities are linked to inputs via
>        ``analysis_process`` entities.

Reviewer: ```suggestion
-  The ``analysis_file`` entities are linked to inputs via ``analysis_process`` entities.
```
Line wrapping and indentation in source causes odd bolding of first line and indentation in displayed output.

Author: Nice catch.

-  Each ``analysis_process`` is described by an ``analysis_protocol``

> -  The ``analysis_protocol.protocol_core.protocol_id`` will be
>        ``optimus_post_processing_v1.0.0`` for stratified, project-level
>        matrices and ``optimus_v1.0.0`` for the intermediate ones

Reviewer: ```suggestion
-  The ``analysis_protocol.protocol_core.protocol_id`` will be ``optimus_post_processing_v1.0.0`` for stratified, project-level matrices and ``optimus_v1.0.0`` for the intermediate ones
```
Line wrapping and indentation in source causes odd bolding of first line and indentation in displayed output.

Any intermediate matrices created during the processing are described as
``analysis_file``, but the ``analysis_file.provenance.submitter_id``
property is omitted.

.. figure:: media/dcp2-matrices.png

>    Graphical interpretation of how DCP/2 generated matrices are
>    linked to the experimental graph

Reviewer: ```suggestion
   Fig. Graphical interpretation of how DCP/2 generated matrices are linked to the experimental graph
```
(nitpick) Text occurs before image in original doc.

Author: This is just the nature of how Github displays rst figures.

DCP/1-generated matrices
------------------------

> -  The ``supplementary_file.provenance.submitter_id`` field is set to
>        ``c9efbb15-c50c-5796-8d15-35e9e1219dc5``. This UUIDv5 was
>        generated using the same namespace UUID as for
>        `Contributor-generated matrices (CGMs)`_ and the name
>        ``dcp1 matrix service``.

Reviewer: ```suggestion
-  The ``supplementary_file.provenance.submitter_id`` field is set to ``c9efbb15-c50c-5796-8d15-35e9e1219dc5``. This UUIDv5 was generated using the same namespace UUID as for `Contributor-generated matrices (CGMs)`_ and the name ``dcp1 matrix service``.
```
Line wrapping and indentation in source causes odd bolding of first line and indentation in displayed output.

in section ....

So Ingest would describe these matrices using an appropriate metadata
graph, similar to how analysis describes their DCP/2-generated matrices.

This would not require any schema changes but the hacky submitter_id
mechanism would have to be used.

TBD: How do we distinguish these organically described CGMs from DCP/2
generated matrices? [Probably via a dedicated submitter ID)

TBD: Do we migrate the current non-organically described CGMs to this
new scheme? If so, who does it and how does this relate to versioning?

TBD: What about intermediate files like BAMs or intermediate matrices?
Should they be required or can the graph skip them and instead use a
simple process to connect the primary sequence data (fastq for example)
to the project-level matrices?

WIP: Data Browser & Azul catalogs
---------------------------------

Migrated from in `another document`_.

.. _another document: https://docs.google.com/document/d/1ok7VpGJJaPP-jqP_9pU3A5KsMdPmRDBXS-KfsNu5-Nc/edit?usp=sharing

We change the Data Browser URL scheme to include an optional initial
version identifier in the URL path

.. parsed-literal::

>     https://data.humancellatlas.org/explore/\ **v2/**\ projects?catalog=dcp1&filter=...

Reviewer: (nitpick) This displays with `https://data.humancellatlas.org/explore/` as a link to `https://data.humancellatlas.org/explore/`. Is there any way to prevent the auto-linking?

Author: Not that I've found :/

.. role:: raw-html(raw)
   :format: html
.. |nn| replace:: :raw-html:`<blockquote>`
.. |ne| replace:: :raw-html:`</blockquote>`

===================
DCP/2 System Design
===================

> |nn| Text in a block quote (like this) is not normative. |ne|

Reviewer 2: ```suggestion
|nn| Text in a block quote (like this) is not normative. It was included for informational purposes only and does not bear on the implementation. |ne|
```
